src/network.cpp

```diff
@@ -0,0 +1,182 @@
+// local
+#include "network.hpp"
+
+// std
+#include <fstream>
+#include <iostream>
+
+Network::Network(const std::vector<uint32_t> &layers_sizes)
+    : m_LayersSizes(layers_sizes) {
+  for (std::size_t l = 1; l < m_LayersSizes.size(); l++) {
+
+    cv::Mat biases = cv::Mat::zeros(layers_sizes[l], 1, CV_64F);
+    cv::randu(biases, 0.0, 1.0);
+    m_Biases.push_back(biases);
+
+    cv::Mat weights =
+        cv::Mat::zeros(layers_sizes[l], layers_sizes[l - 1], CV_64F);
+    const double limit =
+        std::sqrt(1.0 / (layers_sizes[l] + layers_sizes[l - 1]));
+    cv::randu(weights, 0, limit);
+    m_Weights.push_back(weights);
+  }
+}
+
+std::vector<double>
+Network::FeedForward(const std::vector<double> &input) const {
+  if (input.size() != m_LayersSizes[0]) {
+    spdlog::error("[Network::FeedForward]: input size doesn't match "
+                  "m_LayersSizes[0] size\n");
+  }
+
+  cv::Mat activation = cv::Mat(input).reshape(1, input.size());
+
+  for (std::size_t i = 0; i < m_Weights.size(); i++) {
+    cv::Mat weighted_input;
+    cv::gemm(m_Weights[i], activation, 1.0f, m_Biases[i], 1.0f, weighted_input);
+    activation = Sigmoid(weighted_input).clone();
+  }
+
+  std::vector<double> output;
+  activation.reshape(1, 1).copyTo(output);
+  return output;
+}
+
+void Network::SochasticGradientDescent(std::vector<Sample> &training_data,
+                                       std::size_t epochs,
+                                       std::size_t mini_batch_size,
+                                       double learning_rate,
+                                       double regularization_parameter,
+                                       const std::vector<Sample> &test_data) {
+
+  if (regularization_parameter < 0.0) {
+    spdlog::error("[Network::SochasticGradientDescent]: "
+                  "regularization_parameter have to non negative number\n");
+  }
+  if (learning_rate <= 0.0) {
+    spdlog::error("[Network::SochasticGradientDescent]: "
+                  "regularization_parameter have to be a positive number\n");
```

Comment:
I guess you forgot to fix some of the issues. I think you mean to say learning_rate here instead of regularization_parameter.